src/wp-includes/class-wp-block.php

```diff
@@ -191,6 +191,204 @@ public function __get( $name ) {
+			// Get the value based on the source.
+			if ( ! isset( $binding_source['args'] ) ) {
+				$source_args = array();
+			} else {
+				$source_args = $binding_source['args'];
+			}
+			$source_value = call_user_func_array( $source_callback, array( $source_args, $this, $binding_attribute ) );
+			// If the value is null, process next attribute.
+			if ( is_null( $source_value ) ) {
+				continue;
+			}
+
+			// Process the HTML based on the block and the attribute.
+			$modified_block_content = $this->replace_html( $modified_block_content, $this->name, $binding_attribute, $source_value );
+		}
+		return $modified_block_content;
+	}
+
+	/**
+	 * Depending on the block attributes, replace the HTML based on the value returned by the source.
+	 *
+	 * @since 6.5.0
+	 *
+	 * @param string $block_content Block content.
+	 * @param string $block_name The name of the block to process.
+	 * @param string $block_attr The attribute of the block we want to process.
+	 * @param string $source_value The value used to replace the HTML.
+	 */
+	private function replace_html( string $block_content, string $block_name, string $block_attr, string $source_value ) {
+		$block_type = $this->block_type;
+		if ( null === $block_type || ! isset( $block_type->attributes[ $block_attr ] ) ) {
+			return $block_content;
+		}
+
+		// Depending on the attribute source, the processing will be different.
+		switch ( $block_type->attributes[ $block_attr ]['source'] ) {
+			case 'html':
+			case 'rich-text':
+				$block_reader = new WP_HTML_Tag_Processor( $block_content );
+
+				// TODO: Support for CSS selectors whenever they are ready in the HTML API.
+				// In the meantime, support comma-separated selectors by exploding them into an array.
+				$selectors = explode( ',', $block_type->attributes[ $block_attr ]['selector'] );
+				// Add a bookmark to the first tag to be able to iterate over the selectors.
+				$block_reader->next_tag();
+				$block_reader->set_bookmark( 'iterate-selectors' );
+
+				// TODO: This shouldn't be needed when the `set_inner_html` function is ready.
+				// Store the parent tag and its attributes to be able to restore them later in the button.
+				// The button block has a wrapper while the paragraph and heading blocks don't.
+				if ( 'core/button' === $block_name ) {
+					$button_wrapper                 = $block_reader->get_tag();
+					$button_wrapper_attribute_names = $block_reader->get_attribute_names_with_prefix( '' );
+					$button_wrapper_attrs           = array();
+					foreach ( $button_wrapper_attribute_names as $name ) {
+						$button_wrapper_attrs[ $name ] = $block_reader->get_attribute( $name );
+					}
+				}
+
+				foreach ( $selectors as $selector ) {
+					// If the parent tag, or any of its children, matches the selector, replace the HTML.
+					if ( strcasecmp( $block_reader->get_tag( $selector ), $selector ) === 0 || $block_reader->next_tag(
```

Comment:
There is currently an error being triggered in `trunk`:
```
Deprecated: strcasecmp(): Passing null to parameter #1 ($string1) of type string is deprecated in /var/www/html/wp-includes/class-wp-block.php on line 330
```

To reproduce:
1. Create a new post
2. Add a heading block and add the text 'default' to it
3. Using the block settings menu, click 'Create pattern'.
4. In the resulting modal, give the pattern a name and click 'Create'
5. Click the 'Edit original' button that appears on the resulting pattern block
6. Select the heading block and check the 'Allow instance overrides' option in the advanced inspector tools
7. Save and click the 'Back' button from the topbar
8. Now you're back in the post editor, change the heading text 'default' to 'override'

Preview the post
Expected: The preview should show the heading with the text 'override'
Actual: It still shows 'default' and the above error is shown

Reply:
I believe having something like this should solve the issue:

```php
$parent_tag = $block_reader->get_tag();
if ( ! is_null( $parent_tag ) || strcasecmp( $parent_tag, $selector ) === 0 || $block_reader->next_tag(
    array(
	'tag_name' => $selector,
    )
) )
```

Reply:
when using `next_tag()` we will always get a string for `get_tag()` unless no tag was found.

it would be better to check the result of `next_tag()` so we don't descend into this function as if a tag were matched.

```php
if ( ! $block_reader->next_tag() ) {
	continue; // or whatever control flow is appropriate
}
```